wanda/tar_stream.go

```diff
@@ -0,0 +1,104 @@
+package wanda
+
+import (
+	"archive/tar"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"sort"
+	"time"
+)
+
+// tarStream is a stream of files that can be output as a tar stream.
+// It implements io.WriterTo.
+type tarStream struct {
+	files map[string]*tarFile
+
+	// all files will use this mod time default. This makes the stream
+	// deterministic and cachable.
+	modTime time.Time
+}
+
+// newTarStream creates a new tarball stream.
+func newTarStream() *tarStream {
+	return &tarStream{
+		files:   make(map[string]*tarFile),
+		modTime: DefaultTime,
+	}
+}
+
+// DefaultTime is the default timestamp to use in all files for docker input.
+// This makes the build deterministic and cachable.
+var DefaultTime = time.Date(2020, time.January, 1, 0, 0, 0, 0, time.UTC)
+
+// addFile adds a file to the tar stream. If meta is null, it will read the
+// file from the file system to determin the mode, and use the root user as
+// the user and group ID.
+func (s *tarStream) addFile(name string, meta *tarMeta, src string) {
+	s.files[name] = &tarFile{
+		name:    name,
+		srcFile: src,
+		meta:    meta,
+	}
+}
+
+func (s *tarStream) sortedNames() []string {
+	var names []string
+	for name := range s.files {
+		names = append(names, name)
+	}
+	sort.Strings(names)
+	return names
+}
+
+func (s *tarStream) writeTo(tw *tar.Writer) error {
+	names := s.sortedNames()
+
+	for _, name := range names {
+		f := s.files[name]
+		if err := f.writeTo(tw, s.modTime); err != nil {
+			return fmt.Errorf("write file %q to stream", name)
+		}
+	}
+	return nil
+}
+
+// WriteTo writes the entire stream out to w, implements io.WriterTo.
+func (s *tarStream) WriteTo(w io.Writer) (int64, error) {
```

Comment:
name of this function is similar/same with the one above?

Reply:
this one is "public" and implements the interface, the other one is private and does the real thing.